src/pudl/output/ferc1.py

```diff
@@ -1399,3 +1402,341 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
+            child_id = (child.source_table, child.xbrl_factoid)
+            digraph.add_edge(node_id, child_id)
+            # Recursively add all the child nodes and edges:
+            child_graph = child.to_networkx()
+            digraph.add_nodes_from(child_graph.nodes)
+            nx.set_node_attributes(digraph, dict(child_graph.nodes))
+            digraph.add_edges_from(child_graph.edges)
+        return digraph
+
+    def to_leafy_meta(self: Self) -> pd.DataFrame:
+        """Convert an XbrlCalculationTree back into a pandas dataframe.
+
+        - Indexed by (source table, xbrl_factoid)
+        - Each family of tags gets its own column
+        - Weights reflect updated / propagated values
+        - FOREST ONLY: Include column indicating the ID of the root node
+        """
+        G = self.to_networkx()  # noqa: N806
+
+        leaves = [n for n, d in G.out_degree() if d == 0]
+        H = nx.DiGraph()  # noqa: N806
+        H.add_nodes_from(leaves)
+        nx.set_node_attributes(H, dict(G.nodes))
+
+        idx_cols = ["source_table", "xbrl_factoid"]
+        rows = []
+        for node in H.nodes:
+            row = pd.concat(
+                [
+                    pd.DataFrame(
+                        {"source_table": [node[0]], "xbrl_factoid": [node[1]]}
+                    ),
+                    pd.json_normalize(H.nodes(data=True)[node]),
+                ],
+                axis="columns",
+            )
+            rows.append(row)
+        leafy_meta = pd.concat(rows)
+        return leafy_meta.set_index(idx_cols, drop=True).sort_index().convert_dtypes()
+
+
+class XbrlCalculationForestFerc1(BaseModel):
+    """A class representing a collection of :class:`XbrlCalculationTreeFerc1`."""
+
+    trees: list[XbrlCalculationTreeFerc1] = []
+
+    @classmethod
+    def from_exploded_meta(
+        cls,
+        source_tables: list[str],
+        xbrl_factoids: list[str],
+        exploded_meta: pd.DataFrame,
+        propagate_weights: bool = True,
+        tags_df: pd.DataFrame | None = None,
+    ) -> "XbrlCalculationForestFerc1":
+        """Build several :class:`XbrlCalculationTreeFerc1`'s from exploded metadata.
+
+        - Build trees from all of the seeds.
+        - Convert all of the trees into NX graphs.
+        - Add all the nodes from those graphs into a single graph.
+        - Verify that this graph is a forest (one or more trees)
+        - Identify all connected components of this graph.
+        - Verify that the root nodes of each of these trees is one of the seeds.
+        - Convert these Graphs back into Calculation Trees.
+        """
+        trees = [
+            XbrlCalculationTreeFerc1.from_exploded_meta(
+                source_table=source_table,
+                xbrl_factoid=xbrl_factoid,
+                exploded_meta=exploded_meta,
+            )
+            for source_table, xbrl_factoid in zip(source_tables, xbrl_factoids)
+        ]
```

Comment:
By zipping together the two lists, I'm creating a set of tuples that map the elements of the lists together 1:1.  These lists contain all the (probably root) facts that we want to ensure are included in the output tree. They could be a hand-picked set of facts that we care about from, say, the top-level income statement table, or they could be the list of all facts which show up in the income statement table. If not all of them are actually root facts, that's fine -- we use the in-degree and out-degree to figure out which is what later on.

If we give it a non-existent (table, fact) then I think it will interpret that as having gone out of bounds (i.e. it won't find that element in the `exploded_meta` table) -- I was thinking of adding a check that can ensure when this happens the `table` is always not found at all in the index of `exploded_meta` because if the table **is** found, and it's the **fact** that makes it missing... that's probably a typo or something.